### PR TITLE
Pinmux no longer buffers

### PR DESCRIPTION
**Pull request by HU90m**
Removed the `inout`s from the `pinmux` and hooked it up in verilator!

In order to do this in a nice way, I rewrote topgen a little (more) and fixed the data format so some assumptions didn't need to be made. I've tried to keep changes to `pinmux.sv` minimal.

*I had used structs and unions but support was patchy, so big unpacked arrays it is.*

### CODE REVIEW

**Comment by HU90m** [comment]
Currently passing all merged tests bar the I²C test

**Comment by elliotb-lowrisc** [comment]
Synth 8-6104 warning indeed seem to be gone. :+1:

File: doc/ip/pinmux.md
Comment (marnovandermaas) [inline doc/ip/pinmux.md:160]: Hmmm, what about the default values here? I kind of wanted the software to be able to write 2 to all selectors to make things default.

File: rtl/system/sonata_system.sv
Comment (marnovandermaas) [inline rtl/system/sonata_system.sv:1023]: This is now no longer true:
0: general purpose
1: Raspberry Pi
2: Arduino Shield
3: Pmod

**Review by marnovandermaas — COMMENTED** [review]
Some initial comments, but I wasn't able to go through the whole of the last commit. It's a big change.

File: util/top_gen/parser.py
Comment (marnovandermaas) [inline util/top_gen/parser.py:107]: Is there a reason why you are hard-coding this? If we added PWM for example we would have to add it in the toml and also here.

File: dv/verilator/top_verilator.sv
Comment (marnovandermaas) [inline dv/verilator/top_verilator.sv:163]: What does `aux` mean here?

File: dv/verilator/top_verilator.sv
Comment (marnovandermaas) [inline dv/verilator/top_verilator.sv:307]: Although not necessary to make a tri-state signal, it might still be useful to have this visible in simulation.

File: rtl/fpga/top_sonata.sv
Comment (marnovandermaas) [inline rtl/fpga/top_sonata.sv:240]: We tend to use `logic` instead of `wire`. Assigning combinatorially to a `logic` essentially makes it a wire.

File: rtl/fpga/top_sonata.sv
Comment (marnovandermaas) [inline rtl/fpga/top_sonata.sv:325]: Thanks :)

File: util/top_gen/generator.py
Comment (marnovandermaas) [inline util/top_gen/generator.py:174]: Nit: default value above direction.

File: util/top_gen/generator.py
Comment (marnovandermaas) [inline util/top_gen/generator.py:343]: Is this so that we can change the ordering later?

**Review by marnovandermaas — COMMENTED** [review]
I think this PR does what it says. I've left a few comments as I was going through.

This is not something you need to do in this PR, but in the future it would have been nice to have the last commit split up into multiple commits. For example: extracting the buffering from the pinmux, rewriting the parsing and changing the named mapping to the packed array. This is mainly to help reviewing.

File: dv/verilator/top_verilator.sv
Comment (HU90m) [reply dv/verilator/top_verilator.sv:163]: Auxiliary I think, this pre-dates the PR and I was keep changes minimal. I would like to remove this as well.

File: util/top_gen/parser.py
Comment (HU90m) [reply util/top_gen/parser.py:107]: And also to the templates and RTL. I think this is a good sanity check to have.

File: util/top_gen/generator.py
Comment (HU90m) [reply util/top_gen/generator.py:343]: Yep

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [reply rtl/fpga/top_sonata.sv:240]: Three things here:

1. `sonata_out_pins_t` and `sonata_inout_pins_t` are already `logic` from the `typedef`, maybe not hugely surprising Verilog apparently allows you to tag a `wire` on as well (or at least Vivado/Verilator aren't complaining about it) but not something we should actually utilize. So we should not use `wire` with these types as is
2. I'd have a preference for using `wire` in the particular case of nets that can be tri-stated. `logic` can be state holding element which we definitely can't have for tri-state so `wire` is a better statement of intent as well as pointing out there's something a bit unusual about this net (given we basically use `logic` everywhere). This is also clearly stated in our style guide: https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#use-logic-for-synthesis
> There are exceptions for places where logic is inappropriate. For example, nets that connect to bidirectional (inout) ports must be declared with wire. 
3. It may be better to never declare nets like this that can be tri-state at all, instead preferring to confine tri-state nets entirely to module ports that have `'inout` so it's always clear these are tri-state nets. This can be achieved by having a `padring` module that contains the pad interface logic ala opentitan

https://github.com/lowRISC/opentitan/blob/3e3107c7c651612b73d7d7a0004aee7fc5737664/hw/top_earlgrey/rtl/autogen/chip_earlgrey_asic.sv#L307-L753

Note the concatenation syntax that allows you to have a bunch of named inout pins get put into an array you can use within the module.

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [inline rtl/fpga/top_sonata.sv:445]: Isn't this the wrong way round? `ser0_tx` is an output pin and `output_pin[OUT_PIN_SER0_TX]` is the value it should be driven to

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [inline rtl/fpga/top_sonata.sv:435]: If this is intended to drive top-level `output` pins as the name suggests (i.e. specifically not `inout` pins) then it should be not tri-stated. Instead it can just be 0 when output isn't enabled.

There is an argument that we never have any top-level `output` pins and they're all `inout` of course but I think for now given we do have top-level `output` we just drive to 0 when they're not enabled.

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [inline rtl/fpga/top_sonata.sv:438]: Ultimately this is an argument for have specific pad instances that have the top-level IO directly connected through to (as in using exclusively port IO no intermediate `wire` nets that aren't part of a port declaration). Within that pad instance you can do then do whatever tool/technology specific things you need to do. That's what's done in OpenTitan.

**Review by GregAC — COMMENTED** [review]
Thanks for this @HU90m I think this is almost there just the stuff around `logic` v `wire` and exactly how you deal with the tri-stating nets as I've commented on that needs sorting out.

Ultimately I'd like to use a pad wrapper module exactly as we do in OpenTitan but it's not needed for this PR.

Please make sure this gets some testing on FPGA using @AlexJones0's pinmux testing work before it gets merged.

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [reply rtl/fpga/top_sonata.sv:435]: Of course you could define in the pinmux config the default state of an output when it's not connected to anything but I think best to minimise new stuff in this PR.

File: rtl/fpga/top_sonata.sv
Comment (HU90m) [reply rtl/fpga/top_sonata.sv:438]: [I did have a padring module that used the OT pad wrapper](https://github.com/HU90m/sonata-system/commit/414c6d33319a784dd84779f6ca3bb064d492dc84#diff-36f2138a2eccdc3286e62dc9c05fd946567aa9a18831a5e9d3b864339e7fd8e8), but came across this issues.

I agree this would be very nice to have this but as you'd guessed I was trying to keep the changes in this PR as minimal as possible. Definitely do want something like this in the future.

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [reply rtl/fpga/top_sonata.sv:240]: Ah good point, I keep forgetting these details!

In which case `wire` in front of these type should be fine and in line with the style guide but vital we use `wire` (we don't want an implicit `var`).

Though you may run into issues using this nonetheless. Ultimately tri-state drivers and nets do not exist in the FPGA, only at the very edge at the pad instance. You need the synthesis tool to realise you're creating a piece of tri-state logic that can be directly implemented by the tri-state driver available to it (the one in a pad) that's then directly connected out to the pin. It should be able to see this through a wire connect from an IO to another module but a possibility it doesn't (may explain some of the difficulties you had with vivado and using the OT pad wrappers).

File: dv/verilator/top_verilator.sv
Comment (marnovandermaas) [reply dv/verilator/top_verilator.sv:307]: Fair enough

File: dv/verilator/top_verilator.sv
Comment (marnovandermaas) [reply dv/verilator/top_verilator.sv:163]: Ah, didn't know this predates the PR. Resolving this as it is not essential.

File: util/top_gen/parser.py
Comment (marnovandermaas) [reply util/top_gen/parser.py:107]: Ok, I assume you put this in because you ran into issues, so let's resolve this for now.

File: data/pins_sonata.xdc
Comment (GregAC) [inline data/pins_sonata.xdc:94]: Is the `IO_BUFFER_TYPE NONE` required because the pad primitive gives us a an explicit buffer instantiation?

File: rtl/fpga/top_sonata.sv
Comment (GregAC) [reply rtl/fpga/top_sonata.sv:443]: Still need to assign to `0` when disabled here.

File: rtl/fpga/padring.sv
Comment (GregAC) [inline rtl/fpga/padring.sv:19]: First time I've seen this syntax (parameter in square brackets after a module name) but if it works it works :)

**Review by GregAC — APPROVED** [review]
Ok this is looking good from my perspective, thanks @HU90m 

Approving now to unblock but a few things before you merge

- Approval from @marnovandermaas 
- Output default to `0` when not enabled rather than `z`
- Ensure this is tested on a real board use @AlexJones0's pinmux tests.

**Comment by HU90m** [comment]
> * Output default to `0` when not enabled rather than `z`

Ah thank you! I forgot about this

File: rtl/fpga/padring.sv
Comment (HU90m) [reply rtl/fpga/padring.sv:19]: Yeah, I really like module arrays module arrays. I think they are pretty neat and a little more semantic than generator loops.

File: data/pins_sonata.xdc
Comment (HU90m) [reply data/pins_sonata.xdc:94]: Exactamundo, you can put `(* io_buffer_type = "none" *)` attributes on the top level ports in the SystemVerilog, but I thought it was neater having them in the TCL.

**Review by AlexJones0 — COMMENTED** [review]
Can confirm from manual testing that after rebasing #227 and #230 on this branch that both the automated pinmux tests and the manual pinmux muxing check are still passing on FPGA.

**Comment by HU90m** [comment]
> Can confirm from manual testing that after rebasing #227 and #230 on this branch that both the automated pinmux tests and the manual pinmux muxing check are still passing on FPGA.

Yay! thanks Alex

**Review by marnovandermaas — APPROVED** [review]
Let's get this in and iterate over it with follow up PRs.

**Pull request merged**
